### PR TITLE
[DPE-3947] Add relation interface

### PR DESCRIPTION
**Pull request by welpaolo**

### CODE REVIEW

File: src/events/provider.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            )
+        except Exception as e:
+            self.logger.error(e)
+            raise RuntimeError(
+                f"Impossible to create service account: {service_account} in namespace: {namespace}"
+            )
+
+        sa_configuration = {"service-account": service_account, "namespace": namespace}
+
+        # update connection parameters in the relation data bug
```
Comment (deusebio) [inline src/events/provider.py]:
```suggestion
        # update connection parameters in the relation data bag
```

File: src/relations/spark_sa.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+
+
+class IntegratorHubServiceAccountEvents(CharmEvents):
+    """Event descriptor for events raised by ServiceAccountProvider."""
+
+    account_requested = EventSource(ServiceAccountRequestedEvent)
+    account_released = EventSource(ServiceAccountReleasedEvent)
+
+
+class ServiceAccountProvider(Object):
```
Comment (deusebio) [inline src/relations/spark_sa.py]: **issue** It seems that in this implementation we are still using the old pattern for the data platform libs. In particular, in Provider and Requirer we still mix handlers and core (IO with databags). Given the charm is following already the new pattern, I would also try to already use the same pattern for the lib. 

Actually, given that this file is still in the main source code, we could actually import from data platform libs the needed classes/functionalities. Eventually this may very well become a lib, but to start with we can just use the functionalites if that is easier. This way (we would support secrets from the get go)

File: src/events/provider.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        self.sa.update_connection_info(relation_id, sa_configuration)
```
Comment (deusebio) [inline src/events/provider.py]: **question, out-of-scope** If I read the code correctly, it would seem that the requirer and provider databags is actually the same, with for both "service-account" and "namespace". I'm actually wondering whether it is appropriate or if this would generate confusion. Maybe it would be better to have "requested-service-account" (and maybe we can not use "namespace" and use the namespace where we are currently running). 

Also, I'm thinking that maybe the requirer side should also provide its service account, such that the provider could add permission for that service account to get and read the created resources (possibly not modify, just to not mess up with what integration hub is doing, and have one single controlloer for a resource). If that is the case, we could maybe interpret the relation as requirer is asking to create some K8s resources (service account and secrets) and provider replies with the list of resources, but also making sure that permission are set accordingly. Either way, this is food for thoughts, and out of scope here.

File: src/events/provider.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    """Class implementing Spark Service Account Integration event hooks."""
+
+    def __init__(self, charm: CharmBase, context: Context, workload: ConfigurationHubWorkloadBase):
+        super().__init__(charm, "service-account")
+
+        self.charm = charm
+        self.context = context
+        self.workload = workload
+
+        self.configuration_hub = ConfigurationHubManager(self.workload)
```
Comment (deusebio) [inline src/events/provider.py]: **question** This is weird, why is this needed? Do we need to restart the service or update configurations?

File: tests/integration/test_provider.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+    )
+
+    await ops_test.model.wait_for_idle(apps=[APP_NAME, DUMMY_APP_NAME], timeout=600)
+
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(apps=[APP_NAME, DUMMY_APP_NAME], status="active")
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest, namespace):
```
Comment (deusebio) [inline tests/integration/test_provider.py]:
```suggestion
async def test_service_account_relations(ops_test: OpsTest, namespace):
```

it seems that we have already build and deployed before :D

File: src/events/provider.py (hunk 3 above)
Comment (deusebio) [reply src/events/provider.py]: **suggestion** I would generally not use `update_connection_info`, but prefer explicit settings like `set_service_account`, a bit like [here](https://github.com/canonical/kafka-operator/blob/main/src/events/provider.py#L95). This can be a bit more typed, safe and clear

**Review by deusebio — CHANGES_REQUESTED** [review]
I believe we need to think a bit the relation through. Re-working it is outside the scope of the PR, but I believe we need to give it a bit more thoughts, on what should be the interface behaviour. 

But beside this, I believe it would be fair to try to already create the correct skeleton for the interface. For this, it would be nice to already have the proper abstractions, meaning a decomposition between data and events, in the libs. The lib here still follow the old pattern, and this I believe this will create (and to some extent is already creating) friction very soon. So maybe it is worth doing it properly from the very beginning. 

In particular, I believe it would be good to have in the context a view of all related applications, like a property like 

```
    @property
    def clients(self) -> List[ServiceAccount]
```

or similar. This will force us to have a databag representation which is disconneted from events handers. the databag representation is generally used for IO operation. If we inherit from dataplatform libs, we could get secrets handling already built in. Also, I would generally be setting the property in the databag being more explicit and using the data object

```
self.context.get_client(relation_id).set_service_account(...) 
```

or something similar, as io operation may be a responsibility of the data object and be more safe. But this is a minor point as - the way they are constructed - both Data objects and Event handlers share the "set_property" method, being the former inherited by the latter.

File: src/events/provider.py (hunk 4 above)
Comment (theoctober19th) [reply src/events/provider.py]: Slightly unrelated question: aren't we calling this Integration Hub instead of Configuration Hub going forward?

File: src/constants.py [hunk 6]
```diff
@@ -13,3 +13,4 @@
 
 # integrations
 S3 = "s3-credentials"
+SERVICE_ACCOUNT = "spark-service-account"
```
Comment (theoctober19th) [inline src/constants.py]: nit: maybe SERVICE_ACCOUNT_RELATION_NAME or something similar would be more appropriate name since SERVICE_ACCOUNT on the first glance looks like it is name of a K8s service account.

**Review by theoctober19th — APPROVED** [review]
Most suggestions / feedbacks are already addressed by Enrico. I've added a few non-blocking comments. Otherwise, it looks good to me.

File: src/core/context.py [hunk 7]
```diff
@@ -66,6 +67,25 @@ def pushgateway(self) -> PushGatewayInfo | None:
...
+
+        Returns:
+            List of service accounts/namespaces managed by the Integration Hub
+        """
+        service_accounts = []
+        for relation in self.client_relations:
+            if not relation or not relation.app:
+                continue
+            service_accounts.append(ServiceAccount(relation, relation.app))
+        return service_accounts
```
Comment (deusebio) [inline src/core/context.py]: **suggestion, niptick** What about a functional approacht?

```
return [ServiceAccount(relation, relation.app) for relation in self.client_relations if if not relation or not relation.app]
```

File: src/events/provider.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            return
+        relation_id = event.relation.id
+
+        service_account = event.service_account
+        namespace = event.namespace
+        self.logger.debug(
+            f"Desired service account name: {service_account} in namespace: {namespace}"
+        )
+        assert service_account is not None
+        assert namespace is not None
```
Comment (deusebio) [inline src/events/provider.py]: **todo** this would raise an exception, and I'm afraid we would just keep retriggering this. Can we think of a better handling? Like move to blocked when the relation information is not appropriate, and effectively do nothing, just return here?

File: src/events/provider.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        if not self.charm.unit.is_leader():
+            return
+
+        service_account = event.service_account
+        namespace = event.namespace
+        self.logger.debug(
+            f"The service account name: {service_account} in namespace: {namespace} should be deleted"
+        )
+        assert service_account is not None
+        assert namespace is not None
```
Comment (deusebio) [inline src/events/provider.py]: **todo** this would never happen if we were careful. Also more worringly, I suppose this would even block the removal of a relation with not appropriate databag

File: src/relations/spark_sa.py [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+    """Event descriptor for events raised by the Requires."""
+
+    account_granted = EventSource(ServiceAccountGrantedEvent)
+    account_gone = EventSource(ServiceAccountGoneEvent)
+
+
+# Integration Hub Provides and Requires
+
+
+class IntegrationHubProvidesData(ProviderData):
```
Comment (deusebio) [inline src/relations/spark_sa.py]:
```suggestion
class IntegrationHubProviderData(ProviderData):
```

Can you fix the spelling similarly to my PR on the DP Libs?

File: src/relations/spark_sa.py
Comment (deusebio) [inline src/relations/spark_sa.py:1]: **praise** Nice! I like it now! Thanks!

File: src/core/domain.py [hunk 11]
```diff
@@ -104,3 +104,20 @@ def endpoint(self) -> str | None:
                 url = data["url"]
                 return url.replace("https://", "").replace("http://", "")
         return None
+
+
+class ServiceAccount(StateBase):
```
Comment (deusebio) [inline src/core/domain.py:109]: **question** Is this going to stay or will this go away in the mid term? Just wondering why we are not using the `IntegrationHubProviderData` instead? The two classes seem to have a very similar scope

**Review by deusebio — COMMENTED** [review]
I have some minor comments, but - also based on the experience with deploying the Spark bundle - I would avoid raising errors on the relation changed events. If we would like this PR to land sooner rather than later, it is also fine to start merging this PR to unblock other tasks, but I would like this to be addressed in the short term

Let me know your prefernce and I'll change the feedback (approval) accordingly

**Review by deusebio — APPROVED** [review]
